service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java

```diff
@@ -71,35 +78,40 @@
   }
 
   @Override
-  public Long saveState(StateDTO state, Long applicationId) {
+  public Long saveState(StateHistoryDTO state, Long applicationId) {
     Assert.notNull(state);
     Assert.notNull(applicationId);
 
-    StateEntity stateEntity = modelMapper.map(state, StateEntity.class);
-
-    stateEntity.setCreationDate(new Date());
-    stateEntity.setApplicationEntity(applicationsRepository.findOne(applicationId));
+    StatesHistoryEntity statesHistoryEntity = modelMapper.map(state, StatesHistoryEntity.class);
+    statesHistoryEntity.setStatesEntity(statesReporitory.findOne(state.getStateDTO().getId()));
+    statesHistoryEntity.setCreationDate(new Date());
+    statesHistoryEntity.setApplicationEntity(applicationsRepository.findOne(applicationId));
 
-    return statesRepository.save(stateEntity).getId();
+    return statesHistoryRepository.save(statesHistoryEntity).getId();
   }
 
 
   @Override
-  public List<StateViewDTO> getStatesByApplicationId(Long applicationId) {
+  public List<StateViewHistoryDTO> getStatesByApplicationId(Long applicationId) {
     Assert.notNull(applicationId);
     ApplicationEntity applicationEntity = applicationsRepository.findOne(applicationId);
 
     Assert.notNull(applicationEntity);
-    List<StateEntity>
+    List<StatesHistoryEntity>
         stateEntities =
-        statesRepository.findByApplicationEntityOrderByStateIndexDesc(applicationEntity);
-    List<StateViewDTO> stateDTOs = modelMapper.map(stateEntities, STATEVIEWDTO_LIST_TYPE);
+        statesHistoryRepository.findByApplicationEntityOrderByCreationDateDesc(applicationEntity);
+    List<StateViewHistoryDTO> stateDTOs = modelMapper.map(stateEntities, STATEVIEWDTO_LIST_TYPE);
 
     SimpleDateFormat simpleDateFormat = new SimpleDateFormat(DATE_FORMAT_CONSTANT);
 
     for (int i = 0; i < stateDTOs.size(); i++) {
```

Comment:
As we don't know which implementation is used for the List interface here, the get(i) operations can have performance penalty. Switch to foreach instead.